src/runtime/omp/omp_instrumentation.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * This file is part of hipSYCL, a SYCL implementation based on CUDA/HIP
+ *
+ * Copyright (c) 2019-2020 Aksel Alpay and contributors
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "hipSYCL/runtime/omp/omp_instrumentation.hpp"
+#include "hipSYCL/runtime/error.hpp"
+#include "hipSYCL/runtime/instrumentation.hpp"
+
+#include <cassert>
+
+namespace hipsycl {
+namespace rt {
+
+void omp_execution_start_timestamp::wait() const { _signal.wait(); }
+
+profiler_clock::time_point
+omp_execution_start_timestamp::get_time_point() const {
+  return _time;
+}
+
+void omp_execution_start_timestamp::record_time() {
+  _time = profiler_clock::now();
```

Comment:
Do I understand correctly that you say that we cannot actually rely on `signal::wait()` to synchronize with `signal::signal()`? The whole point of the `signal_channel` class was to allow for such patterns, if it doesn't do that then I think we should rather fix `signal_channel`. 
Especially since it's not a new class: It has existed before and is e.g. used to implement events in the OpenMP backend (this PR only moves it to a separate header file to make it usable by the instrumentation API). So, if it's broken existing code relying on it will be broken as well.

Reply:
I dug into the standard and *[futures.state]* does say that `shared_future::wait()` synchronizes-with `promise::set_value()`, so the use in `omp_node_event` is definitely file. The issue that remains IMO is that `omp_execution_*_timestamp` can be easily mis-used: Calling `record_time()` twice or calling `get_time_point()` before `wait()` will usually cause a data race because `_time` is not synchronized otherwise. The straight-forward solution should be to transport a value through `signal_channel` instead of the implicit `bool` it uses now.

Reply:
Thanks for checking with the standard!

So the issue is that it could be misused. Calling `get_time_point()` before `wait()` cannot really happen because the `instrumentation_set` is deliberately designed to not grant direct access to instrumentations, and internally executing `wait()` before accessing results. This simplifies the implementation of the instrumentations, because such synchronization issues are now handled at a central place.
In this design, `instrumentation` objects should mainly be seen as a simple vehicle to store a result.

I have now made `record_time()` private such that it can only be accessed by the OpenMP instrumentation setup logic. It is after all more of an implementation detail of OpenMP instrumentation setup. There are now also asserts to check that it was not invoked before.